### ISSUE TITLE
What is missing before publishing Crafty.js 0.6.0?

### ISSUE THREAD

**Issue by kevinsimper** [craftyjs/Crafty#575]
It has been 5 months since Crafty.js has been released, and a lot of the errors we get from the users has already been solved in the development branch. I think the users will really be glad for the changes, especially the ones @starwed has made :+1: 

Also from the outside it seems to people that Crafty.js is no longer under development.
http://html5gameengine.com/details/3/crafty

What do we miss before we can publish version 0.6.0?
- [x] Land desired features 
- [x] Compile [release notes](https://github.com/craftyjs/Crafty/wiki/Release-note-0.6.0)
- [x] Ensure the new documentation is ready to go live (@kevinsimper has access to the site, I think?)
- [x] Run the unit tests in IE9 and IE10.  Checked: Firefox, Safari, Chrome, PhantomJS. 
- [x] Get [public feedback](https://groups.google.com/forum/?fromgroups#!topic/craftyjs/Eq4TClc6WCQ)
- [x] Build Crafty.js and put the compiled files in /dist folder for Bower support

**Comment by potomak** [comment]
I suggest creating a milestone associated with the release of version 0.6.0 collecting all issues that must be closed before tagging the new version.

**Comment by kevinsimper** [comment]
I have created a milestone now, which marks the completeness of version 0.6.0.

@potomak Any issues you think that should be in that milestone?

**Comment by potomak** [comment]
Sorry but I can't help @kevinsimper, I'm involved in this community since a short time.

**Comment by starwed** [comment]
The main thing I want to land before 0.6 is a large change to `animation.js`.  

Currently all animations and tweens run in a kind of tick based way.  I have an almost finished patch that switches this to a time based method, that will allow for more flexible game loops.

It also changes the "SpriteAnimation" interface pretty substantially, so I really want to land it before the next official release.  (We already changed the interface once in the alphas.)  All the functionality is there now, I was just writing tests and docs.

**Comment by kevinsimper** [comment]
That sounds great, and a good idea to have in the 0.6 release.

**Comment by kevzettler** [comment]
@starwed do you feel like pushing your branch or no? Asking incase you think it would be valuable to get some more eyes on it and maybe help contribute.

**Comment by starwed** [comment]
I opened #583 for the branch I've been working on.

**Comment by kevinsimper** [comment]
#583 has been merged.

We should tag it now for release 0.6 :)

**Comment by starwed** [comment]
Yeah, I found a silly bug that was missed by the unit tests and submitted a fix for.

I'll land that and a couple of other small fixes, and cut a testing branch before we do an actual release.

It might be nice to have #499 and #508 (or equivalent) before the actual release, but we can always merge those in with the testing branch.

**Comment by starwed** [comment]
Put a list of things to finish before the actual release.  Oh, and I guess we need to bump the version number too.  :)

**Comment by starwed** [comment]
Oh, let me know if there's anything missing from the [release notes](https://github.com/craftyjs/Crafty/wiki/Release-note-0.6.0), especially if something should go in the major/notable section.

**Comment by starwed** [comment]
The `stage.html` unit tests don't pass in Firefox.   In particular, the `centerOn` test fails.  I'll investigate why.

If someone could run the tests in Safari, IE9 and 10 that would be good!  PhantomJS uses Chrome's V8, but it would still be nice to here about Chrome.

**Comment by starwed** [comment]
Tested in Chrome (31.0.1650.57) and Safari (6.05).

The `stage.html` test fails the same way in all browsers, but passes in PhantomJS.

In Chrome, the matrix inversion test (30) in `math.html` fails.  It passes in Safari, Firefox, and PhantomJS.  It seems to be relying on a particular value for rounding, so that's probably a test problem rather than a real problem in Chrome.  Probably the matrix `equals` function should round to 10 decimal places or something?

**Comment by starwed** [comment]
Ah, it passed in phantom because of a difference in the size of the viewport.  Phantom defaults to 300x400, which happens to avoid the errors I was getting in other browsers.  Probably that test suite should both explicitly set the viewport size, and run the tests in a couple of variations.

There's one bug in the tests: the values compared to need to be floored, because viewport coordinates are.

There's another type of failure I'm still investigating, though.

**Comment by starwed** [comment]
The other failure I was seeing was due to `_clamp`.  The `centerOn` function calls `pan`, and pan clamps the viewport to the region with entities unless you set the flag otherwise.

So this is just a test problem: the values need to be floored, and the flag needs to be set.  I'll open a PR for that.

_e:_ Fixed in #623

**Comment by starwed** [comment]
Ok, after getting some public testing, there are a few issues that needed fixing.  (And one that still does.)

I tagged the PRs we'll want to backport to the testing branch with a [0.6] label.  There have been some other changes that can wait til the next release; we should just pull in things that either fix regressions or totally broken functionality. 

@kevinsimper Will it be easy to update the docs when we actually do a release?

**Comment by starwed** [comment]
Anyone able to run the unit tests directly in IE9 or IE10?

Maybe @sbyrnes123?

**Comment by starwed** [comment]
No one ever chimed in about tests in IE9/10, but I went ahead and push a release candidate that cherry-picks fixes for important issues from the develop branch.   Hopefully that's a reasonable way to do things!  (I could have also just merged each feature branch into testing, I guess.)

The only other thing to wait on is the docs -- @kevinsimper , since you have the permissions for that repo, is there anything special to do there?  Or will you just build the api and then push it?

**Comment by kevinsimper** [comment]
@starwed Sorry for being late, but yeah it is straight forward. Build det docs and move them to the website. I will make a pull request ready on the site right now.

**Comment by kevinsimper** [comment]
Here is the new documentation pull request:

https://github.com/craftyjs/craftyjs.github.com/pull/23

**Comment by kevinsimper** [comment]
I have added a task, that we should add the compiled version of Crafty to a dist folder, so that people can use Bower to get Crafty.js and so it is easier to get the code, instead of you have to get it down and compile it yourself or find it on another website.

**Comment by kevinsimper** [comment]
DOM: Internet Explorer 9 and 11 error, 10's fine:
![image](https://f.cloud.github.com/assets/1126497/1693860/df71f714-5e96-11e3-9f58-482d0cf24788.png)

That is the only error i encounted, when doing all the test in internet explorer 9 - 10 - 11.

**Comment by starwed** [comment]
I think this is just because IE9 doesn't support such transforms in the first place.    The flag we use is `Crafty.support.css3dtransform`, we can just skip these tests if it's false!  :)

I'm guessing IE11 _does_ support the transforms, but we're detecting it incorrectly?

**Comment by kevinsimper** [comment]
Hehe, I was just about writing that is the `Crafty.support.css3dtransform` that caused it to fail.

http://caniuse.com/transforms3d

I think this is way it is okay in IE10:
http://msdn.microsoft.com/en-us/library/ie/hh673529%28v=vs.85%29.aspx#the_ms_transform_style_property

I do not really understand it deeply, but i do not think it is a critical bug.

**Comment by kevinsimper** [comment]
Nice that you cherry picked some commits for the 0.6 release candidate, but don't you want to include the new scene function? That would be nice addition i think :)

**Comment by starwed** [comment]
> I do not really understand it deeply, but i do not think it is a critical bug.

Yup, it's expected behavior, and more of a bug in the tests.

**Comment by starwed** [comment]
Heh, the scene function didn't quite make the cut.  There's always something it's nice to have, but we need to release eventually!  :)  And it could introduce bugs.

I just realised I accidentally merged in some extra commits into testing.  And indeed, one of those commits turned out to have a bug.  I've reverted the merge, cherry-picked the changes I _meant_ to get, and will push to the testing branch.  (The history on testing will be messier than it needed to be, but oh well.)

This shouldn't require any changes to the docs -- the only commit that touched them was just correcting a mistake!

**Comment by kevinsimper** [comment]
Ah, cool! 

I created a new issue for the IE bug so that can be fixed later.

Made a pull request with the nessary things for bower.

Have already pushed the new documentation, but have not updated the front page. Doing that now.

Will you make a release tag for 0.6.0 or should I do it?

**Comment by starwed** [comment]
Let's tag it once it is merged to release; I'll do that in a sec.  Github does it automatically when we use their release interface.

Since Bower uses the tag to find the files it needs, the bower additions will have to be included before we tag.

**Comment by starwed** [comment]
Ok, I made an [official github release](https://github.com/craftyjs/Crafty/releases/tag/0.6.0)!

Maybe the craftyjs.com landing page should point to those downloads?

I'll close this issue once `craftyjs.com` is updated!  :D

**Comment by starwed** [comment]
Bower install isn't working correctly.

While it realises that it needs version 0.6.0, it refuses to install it because that version has a file that specifies it to be version 0.6.

The solution here is to bump the number to 0.6.1 and create a new tag.

I strongly dislike the way bower handles these things.  (Not versioning, but forcing itself into the project workflow.)

**Comment by kevinsimper** [comment]
That is very wierd about bower, created a issue over there, because it should not be nessary to do these hoops.

https://github.com/bower/bower/issues/998

Good idea point towards github's version!

**Comment by starwed** [comment]
Ok, pushed version 0.6.1.  `bower install crafty` now works correctly AFAICT.  (Although it installs everything; maybe it should just install the `/dist` folder?)

@kevinsimper  Probably no need to rebuild the documentation.  Just make sure the main page calls it 0.6.1.

**Comment by kevzettler** [comment]
GREAT WORK :+1:

**Comment by kevinsimper** [comment]
@starwed Yes, maybe we should use the ignore property in bower.json, but I think it is quite silly that you have to maintain that file if you add new files and/or folders. It should be the other way around, that you list only needed files.

Cool about 0.6.1, we can not wait on bower fixing it's shit.

I just remembered that it would maybe be a good idea to have the unminified sourcemap version of Crafty to be in the dist folder. Now it is just the bare browserified and minified verison. Sourcemap is very useful!

**Comment by kevinsimper** [comment]
@starwed The files in the dist folder should have the updated 0.6.1 in the banner.

Done it manually on craftyjs.com

**Comment by starwed** [comment]
Ok, that's done then!

**Comment by kevinsimper** [comment]
@starwed I just saw that angular does it quite different with bower. They have a whole second repo only for bower, but that seems counter productive :/

**Comment by starwed** [comment]
To be honest, a separate repo would be my preferred solution.
